File: py/minsk/analysis/binding/expressions/unary.py
from typing import Type

from minsk.analysis.binding.expression import BoundExpression
from minsk.analysis.binding.kind import BoundNodeKind
from minsk.analysis.binding.operators.unary import BoundUnaryOperator


class BoundUnaryExpression(BoundExpression):
    operator: BoundUnaryOperator
    operand: BoundExpression

    def __init__(self, operator: BoundUnaryOperator, operand: BoundExpression):
        self.operator = operator
        self.operand = operand

    @property
    def kind(self) -> BoundNodeKind:
        return BoundNodeKind.UnaryExpression

    @property
    def ty(self) -> Type:
        return self.operator.result_type

File: cpp/minsk/source/analysis/binding/nodes/expressions/unary.cpp
#include "minsk/analysis/binding/nodes/expressions/unary.hpp"
#include "minsk/analysis/binding/kind.hpp"
#include "minsk/analysis/binding/nodes/expression.hpp"
#include "minsk/analysis/binding/nodes/expressions/unary/operator.hpp"

minsk::analysis::binding::bound_unary_expression::bound_unary_expression(
    const bound_unary_operator *op, std::unique_ptr<bound_expression> operand)
    : m_op(op), m_operand(std::move(operand)) {}

minsk::analysis::binding::bound_node_kind
minsk::analysis::binding::bound_unary_expression::kind() const {
  return bound_node_kind::unary_expression;
}
minsk::runtime::object_kind
minsk::analysis::binding::bound_unary_expression::type() const {
  return m_op->result_type();
}
const minsk::analysis::binding::bound_unary_operator *
minsk::analysis::binding::bound_unary_expression::op() const {
  return m_op;
}
const minsk::analysis::binding::bound_expression *
minsk::analysis::binding::bound_unary_expression::operand() const {
  return m_operand.get();
}
